src/hash.cr

```diff
@@ -118,6 +290,479 @@ class Hash(K, V)
     new(initial_capacity: initial_capacity) { default_value }
   end
 
+  # ===========================================================================
+  # Internal implementation starts
+  # ===========================================================================
+
+  # Maximum number of `indices_size` for which we do a linear scan
+  # (maximum of 16 entries in `@entries`)
+  private MAX_INDICES_SIZE_LINEAR_SCAN = 32
+
+  # Maximum number of `indices_size` for which we can represent `@indices`
+  # as Pointer(UInt8).
+  private MAX_INDICES_BYTESIZE_1 = 256
+
+  # Maximum number of `indices_size` for which we can represent `@indices`
+  # as Pointer(UInt16).
+  private MAX_INDICES_BYTESIZE_2 = 65536
+
+  # Inserts or updates a key-value pair.
+  private def upsert(key, value) : Nil
+    # Empty hash table so only initialize entries for now
+    if @entries.null?
+      @indices_size_pow2 = 3
+      @entries = Pointer(Entry(K, V)).malloc(4)
+    end
+
+    hash = key_hash(key)
+
+    # No indices allocated yet so try to do a linear scan
+    if @indices.null?
+      # Try to do an upsert by doing a linear scan
+      upserted = upsert_linear_scan(key, value, hash)
+      return if upserted
+
+      # If we couldn't upsert it means the table was full
+      # so a resize might have been done.
+      # Now, it could happen that we are still with less than 16 elements
+      # and so `@indices` will be null, in which case we only need to
+      # add the key-value pair at the end of the `@entries` buffer.
+      if @indices.null?
```

Comment:
why not move this conditional block into `upsert_linear_scan`?

Reply:
I think it might be a bit clearer this way: the `@indices.null?` check is done once again here and if it doesn't hold you know the code will continue running and `@indices` won't be null. If that's hidden inside `upsert_linear_scan` it's harder to tell.